src/Reflex/Patch/DMapWithMove.hs

```diff
@@ -56,36 +70,73 @@ instance {-# INCOHERENT #-} (GEq k, EqTag k v) => EqTag k (From k v) where
   eqTagToEq k _ r = eqTagToEq k (Proxy :: Proxy v) (geqToEq k r)
 #endif
 
+-- |Structure describing a particular change to a key, be it inserting a new key (@From_Insert@), updating an existing key (@From_Insert@ again), deleting a
+-- key (@From_Delete@), or moving a key (@From_Move@).
 data From (k :: a -> *) (v :: a -> *) :: a -> * where
   From_Insert :: v a -> From k v a
+  -- ^Insert a new or update an existing key with the given value @v a@
   From_Delete :: From k v a
+  -- ^Delete the existing key
   From_Move :: !(k a) -> From k v a
+  -- ^Move the value from the given key @k a@ to this key. The source key should also have an entry in the patch giving the current key as @_nodeInfo_to@,
+  -- usually but not necessarily with @From_Delete@.
   deriving (Show, Read, Eq, Ord)
 
+-- |Type alias for the "to" part of a 'NodeInfo'. @'ComposeMaybe' ('Just' k)@ means the key is moving to another key, @ComposeMaybe Nothing@ for any other
+-- operation.
 type To = ComposeMaybe
 
-validPatchDMapWithMove :: forall k v. (GCompare k, EqTag k (ComposeMaybe k)) => PatchDMapWithMove k v -> Bool
-validPatchDMapWithMove (PatchDMapWithMove m) = src == srcFromDst
+-- |Test whether a 'PatchDMapWithMove' satisfies its invariants.
+validPatchDMapWithMove :: forall k v. (GCompare k, GEq k, GShow k) => DMap k (NodeInfo k v) -> Bool
+validPatchDMapWithMove = not . null . validPatchDMapWithMove'
+
+-- |Test whether a 'PatchDMapWithMove' satisfies its invariants and give reasons why it doesn't.
+validPatchDMapWithMove' :: forall k v. (GCompare k, GEq k, GShow k) => DMap k (NodeInfo k v) -> [String]
```

Comment:
Let's change the name of this.  Maybe something like `valiationErrors`?  Since it now returns a list of problems (empty list meaning no problems)

Reply:
I went for long and descriptive: `validationErrorsForPatchDMapWithMove`